crates/ruff_linter/src/rules/ruff/rules/if_key_in_dict_del.rs

```diff
@@ -0,0 +1,164 @@
+
+    fn fix_title(&self) -> String {
+        "Replace with `.pop(..., None)`".to_string()
+    }
+}
+
+/// RUF041
+pub(crate) fn if_key_in_dict_del(checker: &mut Checker, stmt: &StmtIf) {
+    let Some((test_dict, test_key)) = extract_dict_and_key_from_test(&stmt.test) else {
+        return;
+    };
+    let Some((del_dict, del_key)) = extract_dict_and_key_from_del(&stmt.body) else {
+        return;
+    };
+
+    if !is_same_key(test_key, del_key) || !is_same_dict(test_dict, del_dict) {
+        return;
+    }
+
+    if !is_known_to_be_of_type_dict(checker.semantic(), test_dict) {
+        return;
+    }
+
+    let dict_expr = checker.locator().slice(test_dict.range());
+    let key_expr = checker.locator().slice(test_key.range());
+
+    let replacement = format!("{dict_expr}.pop({key_expr}, None)");
+    let edit = Edit::range_replacement(replacement, stmt.range);
+    let fix = Fix::safe_edit(edit);
+
+    let diagnostic = Diagnostic::new(IfKeyInDictDel, stmt.range).with_fix(fix);
+
+    checker.diagnostics.push(diagnostic);
+}
+
+fn extract_dict_and_key_from_test(test: &Expr) -> Option<(&Dict, &Key)> {
+    let Expr::Compare(ExprCompare {
+        left,
+        ops,
+        comparators,
+        ..
+    }) = test
+    else {
+        return None;
+    };
+
+    if !matches!(ops.as_ref(), [CmpOp::In]) {
+        return None;
+    }
+
+    let [right] = comparators.as_ref() else {
+        return None;
+    };
+
+    dict_and_key_verified(right, left.as_ref())
+}
+
+fn extract_dict_and_key_from_del(body: &[Stmt]) -> Option<(&Dict, &Key)> {
+    let [Stmt::Delete(StmtDelete { targets, .. })] = body else {
+        return None;
+    };
+    let [Expr::Subscript(ExprSubscript { value, slice, .. })] = &targets[..] else {
+        return None;
+    };
+
+    dict_and_key_verified(value.as_ref(), slice.as_ref())
+}
+
+fn dict_and_key_verified<'d, 'k>(dict: &'d Dict, key: &'k Key) -> Option<(&'d Dict, &'k Key)> {
+    if !key.is_name_expr() && !key.is_literal_expr() {
+        return None;
+    }
+
+    if !dict.is_name_expr() {
+        return None;
+    }
+
+    Some((dict, key))
+}
+
+fn is_same_key(test: &Expr, del: &Expr) -> bool {
+    match (test, del) {
+        (Expr::Name(ExprName { id: test, .. }), Expr::Name(ExprName { id: del, .. })) => {
+            test.as_str() == del.as_str()
+        }
+
+        (Expr::NoneLiteral(..), Expr::NoneLiteral(..)) => true,
+        (Expr::EllipsisLiteral(..), Expr::EllipsisLiteral(..)) => true,
+
+        (Expr::BooleanLiteral(test), Expr::BooleanLiteral(del)) => test.value == del.value,
+        (Expr::NumberLiteral(test), Expr::NumberLiteral(del)) => test.value == del.value,
+
+        (Expr::BytesLiteral(test), Expr::BytesLiteral(del)) => {
+            Iterator::eq(test.value.bytes(), del.value.bytes())
+        }
+
+        (Expr::StringLiteral(test), Expr::StringLiteral(del)) => {
+            Iterator::eq(test.value.chars(), del.value.chars())
+        }
+
+        _ => false,
+    }
```

Comment:
Could we use `ComparableExpr` here, or are there cases where this rule needs to be stricter than `CompareExpr`?